src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+                    asyncRequest.CompleteUser((object)0);
+                }
+                return 0;
+            }
+
+            // Now readBytes is a payload size.
+            readBytes = _SslState.GetRemainingFrameSize(InternalBuffer, readBytes);
+
+            if (readBytes < 0)
+            {
+                throw new IOException(SR.net_frame_read_size);
+            }
+
+            EnsureInternalBufferSize(SecureChannel.ReadHeaderSize, readBytes);
+
+            if (asyncRequest != null) //Async
+            {
+                asyncRequest.SetNextRequest(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes, s_readFrameCallback);
+
+                _Reader.AsyncReadPacket(asyncRequest);
+
+                if (!asyncRequest.MustCompleteSynchronously)
+                {
+                    return 0;
+                }
+                readBytes = asyncRequest.Result;
+            }
+            else //Sync
+            {
+                readBytes = _Reader.ReadPacket(InternalBuffer, SecureChannel.ReadHeaderSize, readBytes);
+            }
+            return ProcessFrameBody(readBytes, buffer, offset, count, asyncRequest);
+        }
+
+        //
+        // readBytes == SSL Data Payload size on input or 0 on EOF
+        //
+        private int ProcessFrameBody(int readBytes, byte[] buffer, int offset, int count, AsyncProtocolRequest asyncRequest)
+        {
+            if (readBytes == 0)
+            {
+                // EOF
+                throw new IOException(SR.net_io_eof);
+            }
+
+            // Set readBytes to total number of received bytes.
+            readBytes += SecureChannel.ReadHeaderSize;
+
+            // Decrypt into internal buffer, change "readBytes" to count now _Decrypted Bytes_.
+            int data_offset = 0;
+
+            Interop.SecurityStatus errorCode = _SslState.DecryptData(InternalBuffer, ref data_offset, ref readBytes);
+
+            if (errorCode != Interop.SecurityStatus.OK)
+            {
+                byte[] extraBuffer = null;
+                if (readBytes != 0)
+                {
+                    extraBuffer = new byte[readBytes];
+                    Buffer.BlockCopy(InternalBuffer, data_offset, extraBuffer, 0, readBytes);
+                }
+                // Reset internal buffer count.
+                DecrementInternalBufferCount(InternalBufferCount);
+                return ProcessReadErrorCode(errorCode, buffer, offset, count, asyncRequest, extraBuffer);
+            }
+
+
+            if (readBytes == 0 && count != 0)
+            {
+                // Read again since remote side has sent encrypted 0 bytes.
+                DecrementInternalBufferCount(InternalBufferCount);
+                return -1;
+            }
+
+            // Decrypted data start from "data_offset" offset, the total count can be shrunk after decryption.
+            EnsureInternalBufferSize(0, data_offset + readBytes);
+            DecrementInternalBufferCount(data_offset);
```

Comment:
Now that I know what both of these do, I have to ask: "Why do we copy the data that we plan to skip?" In particular, if `readBytes` would fit in the pinnable buffer, but `data_offset + readBytes` wouldn't, why heap allocate another array just to skip the first half?